src/develop/blend.c

```diff
@@ -219,6 +220,178 @@ int dt_develop_blendif_init_masking_profile(struct dt_dev_pixelpipe_iop_t *piece
+                        c32 * (src[i - w3 - 2] + src[i - w3 + 2] + src[i - w2 - 3] + src[i - w2 + 3] + src[i + w2 - 3] + src[i + w2 + 3] + src[i + w3 - 2] + src[i + w3 + 2]) +
+                        c31 * (src[i - w3 - 1] + src[i - w3 + 1] + src[i - w1 - 3] + src[i - w1 + 3] + src[i + w1 - 3] + src[i + w1 + 3] + src[i + w3 - 1] + src[i + w3 + 1]) +
+                        c30 * (src[i - w3] + src[i - 3] + src[i + 3] + src[i + w3]) +
+                        c22 * (src[i - w2 - 2] + src[i - w2 + 2] + src[i + w2 - 2] + src[i + w2 + 2]) +
+                        c21 * (src[i - w2 - 1] + src[i - w2 + 1] + src[i - w1 - 2] + src[i - w1 + 2] + src[i + w1 - 2] + src[i + w1 + 2] + src[i + w2 - 1] + src[i + w2 + 1]) +
+                        c20 * (src[i - w2] + src[i - 2] + src[i + 2] + src[i + w2]) +
+                        c11 * (src[i - w1 - 1] + src[i - w1 + 1] + src[i + w1 - 1] + src[i + w1 + 1]) +
+                        c10 * (src[i - w1] + src[i - 1] + src[i + 1] + src[i + w1]) +
+                        c00 * src[i];
+      out[i] = fminf(1.0f, fmaxf(0.0f, val));
+    }
+  }
+}
+static void fill_mask_border(float *mask, const int width, const int height, const int border)
+{
+  if(border <= 0) return;
+  for(int row = border; row < height - border; row++)
+  {
+    const int idx = row * width;
+    for(int i = 0; i < border; i++)
+    {
+      mask[idx + i] = mask[idx + border];
+      mask[idx + width - i - 1] = mask[idx + width - border -1];   
+    }
+  }
+  for(int col = 0; col < width; col++)
+  {
+    const float top = mask[border * width + MIN(width - border - 1, MAX(col, border))];
+    const float bot = mask[(height - border - 1) * width + MIN(width - border - 1, MAX(col, border))];
+    for(int i = 0; i < border; i++)
+    {
+      mask[col + i * width] = top;
+      mask[col + (height - i - 1) * width] = bot;
+    }   
+  }
+} 
+
+void dt_develop_make_contrast_mask(const float *const ivoid, float *mask, const int width, const int height, const float level, const gboolean inverted, const gboolean labspace)
+{
+  if(level == 0.0f) return;
+  gboolean detail = (level > 0.0f);
+  const float threshold = detail ? 0.01f * sqrf(level) : 0.05f * (1.0f - sqrtf(fabs(level)));
+  
+  float *tmp = NULL;
+  float *lum = NULL;
+  tmp = dt_alloc_align_float((size_t) width * height);
+  lum = dt_alloc_align_float((size_t) width * height);
+  if((tmp == NULL) || (lum == NULL))
+  {
+    dt_control_log(_("could not allocate buffer for contrast mask blending"));
+    goto finish;
+  }
+
+#ifdef _OPENMP
+  #pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(lum, ivoid, width, height, labspace) \
+  schedule(simd:static) aligned(lum, ivoid : 64) 
+#endif
+  for(size_t idx =0; idx < (size_t) width * height; idx++)
+  {
+    const float val = labspace ? 0.02f * ivoid[4 * idx] : 0.333333f * (ivoid[4 * idx] + ivoid[4 * idx + 1] + ivoid[4 * idx + 2]);
+    lum[idx] = lab_f(val);
```

Comment:
why converting to Lab ?

Reply:
again i took this from rt code. could use another function of course but that would lead to re-evaluation of other "magic constants" and calculations